### ISSUE TITLE
Question: API Testing with Passport

### ISSUE THREAD

**Issue by mousetree** [angular-fullstack/generator-angular-fullstack#494]
Hi, I'm not sure if this is the right place to post this question but I'm not having any luck elsewhere. I'm new to testing and I'm not having any luck on SO etc. as they lack the context of whats already provided by this generator.

Most of my routes on the API are protected. What I'm trying to do is write test specs that:
- Test endpoints without authorization (that fail)
- Test endpoints with auth (pass)
- Then test specifics/contents on the data returned by each endpoint while authorized.

What I'm having trouble with, is how to write the tests such that:
- The test users are created (user.model.spec.js clears all users)
- Authenticating the user with passport.js (local strategy is used)
- Test endpoints with auth and without auth and the particulars of the endpoint whilst authorized.

Please help, i'm pulling my hair out!

**Comment by wmertens** [comment]
What I don't understand is why the server authentication doesn't also read the token from the session cookie? Then you can browse the API in the browser without having to add `Authorization` headers. If there's a header, use that, if there is a cookie, use that.

The cookie is being added for most of the OAuth strategies anyway, so why not for all?

As you can see, not doing that results in a headaches and extra code on the API consumer side, without security improvements (AFAIK)?

**Comment by chuyik** [comment]
@wmertens IMO if you allow reading token from cookie, this would raise a CSRF attack issue.
Other website could invoke a AJAX request and the token just be sent by browser automatically rather than programmatically when CORS is allowed or a JSONP request triggered.

**Comment by kingcody** [comment]
@chuyik you're correct regarding CSRF attacks and non http only cookies. However the token cookie is not http only as it is read by `auth.service.js` to log the user in automatically.

**Comment by wmertens** [comment]
@chuyik Not sure if that is a problem, since if you allow CORS/JSONP you presumably are happy with the token being used by other sites. Anyway, so maybe that should be configurable then...

**Comment by chuyik** [comment]
@wmertens That's definitely a security problem even not allow CORS.
Assume that your server allows reading token from cookie, and your resource location is at  `GET http://yours.com/users`.

Then your user visits my website, which secretly invokes a GET request with a script tag like:
```
<script src="http://yours.com/users"></script>
```
And the users data are stolen by me since the cookies are sent by the browser and you read token from them.
So I don't even recommend a configuration there.

**Comment by wmertens** [comment]
@chuyik I was under the impression that CORS prevents exactly this? Can you point me to a proof-of-concept implementation or explanation?

**Comment by chuyik** [comment]
@wmertens I've tried to prove me right but wrong. :(
There's no way to get script or iframe content if the tag is not inline. So it's my bad for showing the wrong example.
If it's not CORS or JSONP, it should be fine to read token from cookie.

**Comment by jeshuamaxey** [comment]
I've adapted @chuyik's code to allow the test to specify a role for the user. Useful for testing endpoints which have different permissions

https://gist.github.com/jeshuamaxey/e88a21f802445bf05e18

**Comment by Awk34** [comment]
@jeshuamaxey PR?

**Comment by jeshuamaxey** [comment]
I'm still ironing out some teething problems to do with tests running before the auth request returns a token. Once I have a fully working solution, I'll set up a PR :)

**Comment by jeshuamaxey** [comment]
FYI I've updated my gists with the implementation I've settled on for testing endpoints which require user authentification. They can be viewed here: https://gist.github.com/jeshuamaxey/e88a21f802445bf05e18

Next step is to develop this into a PR. With any luck I'll push something tonight, although will more likely be sometime mid next week.

**Comment by lxibarra** [comment]
Thanks to everyone contributing on this issue it was really helpful.
